lib/ansible/modules/cloud/vmware/vmware_guest_disk.py

```diff
@@ -0,0 +1,604 @@
+                                state='present',
+                                datastore=None,
+                                autoselect_datastore=True,
+                                disk_unit_number=0,
+                                scsi_controller=0)
+            # Check state
+            if 'state' in disk:
+                if disk['state'] not in ['absent', 'present']:
+                    self.module.fail_json(msg="Invalid state provided '%s' for disk index [%s]."
+                                              " State can be either - 'absent', 'present'" % (disk['state'],
+                                                                                              disk_index))
+                else:
+                    current_disk['state'] = disk['state']
+
+            if current_disk['state'] == 'present':
+                # Select datastore or datastore cluster
+                if 'datastore' in disk:
+                    if 'autoselect_datastore' in disk:
+                        self.module.fail_json(msg="Please specify either 'datastore' "
+                                                  "or 'autoselect_datastore' for disk index [%s]" % disk_index)
+
+                    elif 'autoselect_datastore' not in disk:
+                        # Check if given value is datastore or datastore cluster
+                        datastore_name = disk['datastore']
+                        datastore_cluster = find_obj(self.content, [vim.StoragePod], datastore_name)
+                        if datastore_cluster:
+                            # If user specified datastore cluster so get recommended datastore
+                            datastore_name = self.get_recommended_datastore(datastore_cluster_obj=datastore_cluster)
+                        # Check if get_recommended_datastore or user specified datastore exists or not
+                        datastore = find_obj(self.content, [vim.Datastore], datastore_name)
+                        if datastore is None:
+                            self.module.fail_json(msg="Failed to find datastore named '%s' "
+                                                      "in given configuration." % disk['datastore'])
+                        current_disk['datastore'] = datastore
+                        current_disk['autoselect_datastore'] = False
+                elif 'autoselect_datastore' in disk:
+                    # Find datastore which fits requirement
+                    datastores = get_all_objs(self.content, [vim.Datastore])
+                    if not datastores:
+                        self.module.fail_json(msg="Failed to gather information about"
+                                                  " available datastores in given datacenter.")
+                    datastore = None
+                    datastore_freespace = 0
+                    for ds in datastores:
+                        if isinstance(ds, vim.Datastore) and ds.summary.freeSpace > datastore_freespace:
```

Comment:
I may be mistaken but, I think `isinstance(ds, vim.Datastore)` is not necessary since you get the datastores list from `get_all_objs(self.content, [vim.Datastore])` (so all `ds` should be of type `vim.Datastore`).

Reply:
OK.

Reply:
You forget to make the change.